python/pipeline/stack.py

```diff
@@ -977,6 +978,129 @@ def make(self, key):
+        full_mesh_x, full_mesh_y = np.meshgrid(np.arange(width), np.arange(height))
+
+        # Surface z should be below this value
+        z_lim = int(depth / 2)
+        # Mean intensity of the first frame in the slice should be less than this value
+        z_0_upper_threshold = np.percentile(full_stack, 30)
+
+        for x in r_xs:
+            for y in r_ys:
+                stack_slice_medians = np.percentile(full_stack[0:z_lim, y - r:y + r, x - r:x + r], 50, axis=(1, 2))
+                blurred_slice = ndimage.gaussian_filter1d(stack_slice_medians, gaussian_blur_size)
+
+                upper_threshold_value = upper_threshold_percent * (
+                            (blurred_slice.max() - blurred_slice.min()) - blurred_slice.min())
+                upper_threshold_idx = np.where(blurred_slice > upper_threshold_value)[0][0]
+
+                stack_slice_derivative = ndimage.sobel(blurred_slice)
+                surface_z = np.argmax(stack_slice_derivative)
+
+                if ((surface_z < upper_threshold_idx) and (blurred_slice[0] < z_0_upper_threshold) and
+                        (blurred_slice[-1] > 10)):
+                    surface_guess_map.append((surface_z, y, x))
+
+        if len(surface_guess_map) < min_points_allowed:
+            raise Exception(f"Surface calculation could not find enough valid points for {key}. Only {len(surface_guess_map)} detected")
+
+        # Drop the z-values lower than 5th-percentile or greater than 95th-percentile
+        arr = np.array(surface_guess_map)
+        top = np.percentile(arr[:, 0], 95)
+        bot = np.percentile(arr[:, 0], 5)
+        surface_guess_map = arr[np.where(np.logical_and(arr[:, 0] > bot, arr[:, 0] < top))]
+
+        # Guess for initial parameters
+        initial = [1, 1, int(width / 2), int(height / 2), 1]
+
+        popt, pcov = curve_fit(surface_eqn, (surface_guess_map[:, 2], surface_guess_map[:, 1]),
+                               surface_guess_map[:, 0], p0=initial, maxfev=10000, bounds=bounds)
+
+        calculated_surface_map = surface_eqn((full_mesh_x, full_mesh_y), *popt)
+
+        all_sub_fitted_z = np.zeros((num_iterations, height, width))
+        for i in np.arange(num_iterations):
+            indices = np.random.choice(surface_guess_map.shape[0], int(surface_guess_map.shape[0] * ss_percent),
+                                       replace=False)
+            subsample = surface_guess_map[indices]
+            sub_popt, sub_pcov = curve_fit(surface_eqn, (subsample[:, 2], subsample[:, 1]), subsample[:, 0],
+                                           p0=initial, maxfev=10000, bounds=bounds)
+            all_sub_fitted_z[i, :, :] = surface_eqn((full_mesh_x, full_mesh_y), *sub_popt)
+
+        z_min_matrix = np.percentile(all_sub_fitted_z, 5, axis=0)
+        z_max_matrix = np.percentile(all_sub_fitted_z, 95, axis=0)
+
+        surface_key = {**key, 'surface_method_id': 1, 'surface_im': calculated_surface_map,
```

Comment:
surface method is gonna come with the key, no need to send it here. In fact, right at the start of the make you should probably check that surface_method is 1 and throw a NotImplementedError or PipelineException if it's not

Reply:
I added a variable storing all valid method_ids and now raise a PipelineException if the key value is not in that list.